profiles/permissions.py

```diff
@@ -26,10 +26,18 @@ def has_permission(self, request, view):
         return request.user == view._profile.person
 
 
+class OnlyRequestUserOrAdmin(BasePermission):
+    def has_permission(self, request, view):
+        user = request.user
+        if request.method in SAFE_METHODS and user.is_staff:
+            return True
+        return str(request.user.id) == str(request.data.get("user"))
```

Comment:
Permissions MUST NOT deal with `request.data`.

Reply:
Why?

Reply:
I don't insist on rework of this issue, but strongly recommend to consider it in future.

Ideally, `request.data` should be dealt with only through serializers. Doing `if request.data.get(...)` is pretty close to doing `return Response()`. It works, nobody will die, but is not RESTful. 

Permissions define if user has access to some _specific endpoint URL_. It should not depend on the content of payload, filters (query parameters) etc.

Reply:
And how can we check if user.is_staff or obj.person == request.user (for example)?

Reply:
This kind of check is already implemented if we look at `ViewedCompany` view and permissions.

Reply:
> And how can we check if request.user.is_staff or obj.person == request.user (for example)? What is safe method for this?

If you want to check request user against some object — I recommend including the object ID somewhere in URL, retrieving the object in view (through cached property), and then working with this object.

Reply:
@BelousSofiya OK, here is one more point. 

I'm not sure if it work out as a viable exploit of your code in this case, but I believe you'll at least feel the danger. :)

So, in this permission class, you just dive into content of payload. Which is _not validated_ (as at this point it wasn't processed by serializer). You pick some part of this payload, check if it matches current user, and if it does — permission class says 'OK, you can do this action'. Nice and easy, what can go wrong?

But imagine that a bad actor does the action on an object which doesn't actually belongs to him/her. But nothing stops the bad actor to form any payload he wants. So, he puts his ID in `user` field, and this permission class allows him to perform requested action. Funny, isn't it? :)

Now compare this with the approach I suggest. Object ID is part of URL. Object _is retrieved from DB_, and user is compared to the real object's content. 

Btw, object retrieval can be put into something like:

```python
@cached_property
def _my_object(self):
    return get_object_or_404(
        ObjectModel.objects.visible_to(user=self.request.user),
        pk=self.kwargs['object_pk']
    )
```

So, with properly implemented `.visible_to()` manager, code is clean and guarantees immediate returning 404 if someone tries to access the object not available for him.

Reply:
... in addition to `.visible_to()` we can also have `.editable_by()` etc.

Reply:
We have permission for this. User can put not his id in request, but it wouldn't work. But ok, I can delete user.id from request.

Reply:
Thank you for feedback. I didn't change permissions a lot (maybe we'll create a task for this), but I've deleted request.data from permissions and user.id from request